experimental/sql/src/impl_to_iterators.py

```diff
@@ -308,12 +312,41 @@ def match_and_rewrite(self, op: RelImpl.Project, rewriter: PatternRewriter):
 #===------------------------------------------------------------------------===#
 
 
+def get_batch_and_name_list(
+    op: FuncOp) -> Tuple[list[str], list[it.ColumnarBatch]]:
+  batches = []
+  names = []
+  for o in op.body.ops:
+    if isinstance(o, RelImpl.FullTableScanOp):
+      curr_batch = it.ColumnarBatch.get(
+          TupleType([
+              ArrayAttr.from_list([
+                  convert_datatype(e.elt_type) for e in o.result.typ.schema.data
+              ])
+          ]))
+      batches.append(curr_batch)
+      names.append(o.table_name.data)
+
+  return names, batches
+
+
 def impl_to_iterators(ctx: MLContext, query: ModuleOp):
 
+  names, batches = get_batch_and_name_list(query)
+
+  table_mapping = {}
+
   # Wrapping everything into a main function
-  f = FuncOp.from_region("main", [], [],
-                         Region.from_block_list([query.body.detach_block(0)]))
+  body_block = Block.from_arg_types(batches)
+  body_block.add_ops(
+      [query.body.blocks[0].detach_op(o) for o in query.body.blocks[0].ops])
+  query.body.detach_block(0)
+  f = FuncOp.from_region("main", batches, [],
+                         Region.from_block_list([body_block]))
   query.body.add_block(Block.from_ops([f]))
+  # Populating a mapping from table names to BlockArguments
+  for n, b in zip(names, f.body.blocks[0].args):
+    table_mapping[n] = b
```

Comment:
Does this work if the same table is scanned twice? It'll appear twice in `names` and `batches` but only once in `table_mapping`. Maybe it *does* but it's a bit obscure...

How about having `get_batch_and_name_list` return an (ordered?) dictionary instead?

Reply:
I think it makes most sense to only load each table once, so only have it once in the returned lists. I added a check that the given table has not been loaded previously.